### DISCUSSION TITLE
📣 2020-08-31 Update: improved configuration

### DISCUSSION THREAD

**Discussion by AdamWagner** [AdamWagner/stackline#33] [General]
Configuration is documented in the wiki:
https://github.com/AdamWagner/stackline/wiki/Configuring-stackline

---

From @alin23:

>Another thing that would be great to have is configurable border radius (if it isn't already). width / 3 is a bit too rounded for the Big Sur icons

**Comment by AdamWagner** [comment]
This is a big one! Much-improved configurability is the primary new feature. You can try it out now on the https://github.com/AdamWagner/stackline/tree/feature/configuration branch, or wait for it be merged.

There are 3 configuration themes:

**Appearance**
Customize indicator sizing, coloring, positioning, roundness, animation durations, form factor, etc

**Features**
Enable/disable fuzzy frame detection, click-to-focus, the hacky workaround for https://github.com/Hammerspoon/hammerspoon/issues/2400, or – coming soon – _window titles_.
You'll no longer have to put up with bugs in a feature that you don't even use ;-)

**Paths**
An easier way to specify paths to binaries & scripts without mucking around in the `stackline` source.

---

**Config validation & live-tweaking**

All config values can be edited while `stackline` is running (this should seem familiar from `yabai`).

Values  can be set using the `hs` cli tool:

```sh
hs -c "stackline.config:set('appearance.radius', 3)"
```

… or the ipc port:

```sh
echo ":appearance.radius:3` | hs -m stackline-config
```

All config fields support  `get(key)`, `set(key, val)`, and `getOrSet(key, [val])`. 
In addition, boolean fields support `toggle()`

You can get and set nested config fields via a dot-separated path, like this: `features.fzyFrameDetection.enabled`

Config fields are typed and validated on initialization and when live-set. 
`stackline` will notify you If you can't remember the exact path to a config variable:

![image](https://user-images.githubusercontent.com/1683979/95018114-68102a80-0612-11eb-9546-9ab3f44622a5.png)

---

**The default config**


```lua
c = {}
c.paths = {}
c.appearance = {}
c.features = {}

-- Paths
c.paths.getStackIdxs                  = hs.configdir .. '/stackline/bin/yabai-get-stack-idx'
c.paths.jq                            = '/usr/local/bin/jq'
c.paths.yabai                         = '/usr/local/bin/yabai'

-- Appearance
c.appearance.color                    = { white = 0.90 }
c.appearance.alpha                    = 1
c.appearance.dimmer                   = 2.5                 -- larger numbers increase contrast b/n focused & unfocused state
c.appearance.iconDimmer               = 1.1                 -- custom dimmer for icon
c.appearance.showIcons                = true
c.appearance.size                     = 32
c.appearance.radius                   = 3
c.appearance.padding                  = 4
c.appearance.iconPadding              = 4
c.appearance.pillThinness             = 6

c.appearance.vertSpacing              = 1.2

c.appearance.offset                   = {}
c.appearance.offset.y                 = 2
c.appearance.offset.x                 = 4

c.appearance.shouldFade               = true
c.appearance.fadeDuration             = 0.2

-- Features
c.features.clickToFocus               = true
c.features.hsBugWorkaround            = true

c.features.fzyFrameDetect             = {}
c.features.fzyFrameDetect.enabled     = true
c.features.fzyFrameDetect.fuzzFactor  = 15                  -- window frame dimensions will be rounded to nearest fuzzFactor

c.features.winTitles                  = 'not_implemented'    -- false, true, 'when_switching', 'not_implemented'
c.features.dynamicLuminosity          = 'not_implemented'    -- false, true, 'not_implemented'
```

You can try it out now on the https://github.com/AdamWagner/stackline/tree/feature/configuration branch, or wait for it be merged.

**Comment by alin23** [comment]
This is awesome! You've put a lot of work into this, I can't wait to try it when I get back to my laptop!

**Comment by AdamWagner** [comment]
**init.lua before**

```lua
stackline = require "stackline.stackline.stackline"
```

**init.lua after**

```lua
stackline = require "stackline.stackline.stackline"
stackline:init() -- must initialize stackline in your init.lua file!
```

The changes, of course, allow you to pass in custom config options ;-) 


```lua
stackline = require "stackline.stackline.stackline"

local myStackline = {
    appearance = { 
      showIcons = false,       -- default is true
    },
    features = {
        clickToFocus = false,  -- default is true
        fzyFrameDetect = {
            fuzzFactor = 25    -- default is 30
        },
    },
}

stackline:init(myStackline)  -- ah yea!          
```
